notifications/manager.go

```diff
@@ -0,0 +1,193 @@
+	NotificationTypeReservation
+)
+
+type NotificationsClient interface {
+	SubscribeNotifications(ctx context.Context,
+		in *swapserverrpc.SubscribeNotificationsRequest,
+		opts ...grpc.CallOption) (
+		swapserverrpc.SwapServer_SubscribeNotificationsClient, error)
+}
+
+// Config contains all the services that the notification manager needs to
+// operate.
+type Config struct {
+	// Client is the client used to communicate with the swap server.
+	Client NotificationsClient
+
+	// FetchL402 is the function used to fetch the l402 token.
+	FetchL402 func(context.Context) error
+}
+
+// Manager is a manager for notifications that the swap server sends to the
+// client.
+type Manager struct {
+	cfg *Config
+
+	hasL402 bool
+
+	subscribers map[NotificationType][]subscriber
+	sync.Mutex
+}
+
+// NewManager creates a new notification manager.
+func NewManager(cfg *Config) *Manager {
+	return &Manager{
+		cfg:         cfg,
+		subscribers: make(map[NotificationType][]subscriber),
+	}
+}
+
+type subscriber struct {
+	subCtx   context.Context
+	recvChan interface{}
+}
+
+// SubscribeReservations subscribes to the reservation notifications.
+func (m *Manager) SubscribeReservations(ctx context.Context,
+) <-chan *swapserverrpc.ServerReservationNotification {
+
+	// We'll create a channel that we'll use to send the notifications to the
+	// caller.
+	notifChan := make(
+		chan *swapserverrpc.ServerReservationNotification, //nolint:lll
+	)
+	sub := subscriber{
+		subCtx:   ctx,
+		recvChan: notifChan,
+	}
+
+	m.Lock()
+	m.subscribers[NotificationTypeReservation] = append(
+		m.subscribers[NotificationTypeReservation],
+		sub,
+	)
+	m.Unlock()
+
+	return notifChan
+}
+
+// Run starts the notification manager.
+func (n *Manager) Run(ctx context.Context, readyChan chan<- struct{}) error {
+	// In order to create a valid l402 we first are going to call
+	// the FetchL402 method. As a client might not have outbound capacity
+	// yet, we'll retry until we get a valid response.
+	if !n.hasL402 {
+		n.fetchL402(ctx)
+	}
+
+	var closedChan bool
+
+	// Start the notification runloop.
+	for {
+		// Return if the context has been canceled.
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+
+		callCtx, cancel := context.WithCancel(ctx)
+		notifStream, err := n.cfg.Client.SubscribeNotifications(
+			callCtx, &swapserverrpc.SubscribeNotificationsRequest{},
+		)
+		if err != nil {
+			cancel()
+			log.Errorf("Error subscribing to notifications: %v", err)
+			continue
+		}
+		if !closedChan {
+			close(readyChan)
+		}
+
+		log.Debugf("Successfully subscribed to server notifications")
+
+		for {
+			notification, err := notifStream.Recv()
+			if err == nil && notification != nil {
+				log.Debugf("Received notification: %v", notification)
+				n.handleNotification(notification)
+				continue
+			}
+			log.Errorf("Error receiving "+
+				"notification: %v", err)
+
+			cancel()
+			// Wait for a bit before reconnecting.
+			<-time.After(time.Second * 10)
```

Comment:
I propose this:
```go
select {
case <-time.After(time.Second * 10):
case <-ctx.Done():
}
```
not to keep this goroutine running for 10 seconds after main context is cancelled.